File: 09_TOOLS/longhun-memory/longhun_memory/sm_crypto.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# CONFIRM: #CONFIRM🌌9622-ONLY-ONCE🧬LK9X-772Z
# DNA: #龍芯⚡️丙午·乙未·壬子·丙午·䷙大畜-MEMORY-SM-CRYPTO-v1.0
# License: MulanPSL v2
"""
國密密码模块 · SM3 哈希 + SM4 加密 + SM3 哈希链
═══════════════════════════════════════════════════════
纯 Python 实现，零外部依赖。
来源: CNSH 国密工具模块（cnsh_guomi.py）
标准: GB/T 32905-2016 (SM3) · GB/T 32907-2016 (SM4)
"""

import os
import struct
import time
from typing import Union, List, Optional


# ════════════════════════════════════════════════════
# SM3 哈希算法 (GB/T 32905-2016)
# ════════════════════════════════════════════════════

class SM3:
    """SM3 密码杂凑算法，输出 256 位摘要。"""

    IV = [
        0x7380166F, 0x4914B2B9, 0x172442D7, 0xDA8A0600,
        0xA96F30BC, 0x163138AA, 0xE38DEE4D, 0xB0FB0E4E,
    ]
    T0 = 0x79CC4519
    T1 = 0x7A879D8A

    @staticmethod
    def _rotl(x: int, n: int) -> int:
        return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF

    @staticmethod
    def _ff(x: int, y: int, z: int, j: int) -> int:
        return x ^ y ^ z if j < 16 else (x & y) | (x & z) | (y & z)

    @staticmethod
    def _gg(x: int, y: int, z: int, j: int) -> int:
        return x ^ y ^ z if j < 16 else (x & y) | (~x & z)

    @staticmethod
    def _p0(x: int) -> int:
        return x ^ SM3._rotl(x, 9) ^ SM3._rotl(x, 17)

    @staticmethod
    def _p1(x: int) -> int:
        return x ^ SM3._rotl(x, 15) ^ SM3._rotl(x, 23)

    @classmethod
    def hash(cls, data: Union[str, bytes]) -> bytes:
        """计算 SM3 哈希，返回 32 字节"""
        if isinstance(data, str):
            data = data.encode("utf-8")
        bit_len = len(data) * 8
        data += b"\x80"
        while (len(data) * 8) % 512 != 448:
            data += b"\x00"
        data += struct.pack(">Q", bit_len)

        v = cls.IV[:]
        for i in range(0, len(data), 64):
            block = data[i:i + 64]
            w = [0] * 68
            w[0:16] = [int.from_bytes(block[j:j + 4], "big") for j in range(0, 64, 4)]
            for j in range(16, 68):
                w[j] = cls._p1(w[j - 16] ^ w[j - 9] ^ cls._rotl(w[j - 3], 15)) ^ \
                       cls._rotl(w[j - 13], 7) ^ w[j - 6]
            w1 = [w[j] ^ w[j + 4] for j in range(64)]
            a, b, c, d, e, f, g, h = v
            for j in range(64):
                tj = cls.T0 if j < 16 else cls.T1
                ss1 = cls._rotl((cls._rotl(a, 12) + e + cls._rotl(tj, j % 32)) & 0xFFFFFFFF, 7)
                ss2 = ss1 ^ cls._rotl(a, 12)
                tt1 = (cls._ff(a, b, c, j) + d + ss2 + w1[j]) & 0xFFFFFFFF
                tt2 = (cls._gg(e, f, g, j) + h + ss1 + w[j]) & 0xFFFFFFFF
                d, c, b, a = c, cls._rotl(b, 9), a, tt1
                h, g, f, e = g, cls._rotl(f, 19), e, cls._p0(tt2)
            for idx in range(8):
                v[idx] ^= [a, b, c, d, e, f, g, h][idx]

        return b"".join(x.to_bytes(4, "big") for x in v)

    @classmethod
    def hex(cls, data: Union[str, bytes]) -> str:
        """SM3 哈希 → 十六进制字符串"""
        return cls.hash(data).hex()


# ════════════════════════════════════════════════════
# SM4 分组加密算法 (GB/T 32907-2016)
# ════════════════════════════════════════════════════

class SM4:
    """SM4 分组密码，128 位密钥，128 位分组，32 轮。"""

    FK = [0xA3B1BAC6, 0x56AA3350, 0x677D9197, 0xB27022DC]
    CK = [
        0x00070E15, 0x1C232A31, 0x383F464D, 0x545B6269,
        0x70777E85, 0x8C939AA1, 0xA8AFB6BD, 0xC4CBD2D9,
        0xE0E7EEF5, 0xFC030A11, 0x181F262D, 0x343B4249,
        0x50575E65, 0x6C737A81, 0x888F969D, 0xA4ABB2B9,
        0xC0C7CED5, 0xDCE3EAF1, 0xF8FF060D, 0x141B2229,
        0x30373E45, 0x4C535A61, 0x686F767D, 0x848B9299,
        0xA0A7AEB5, 0xBCC3CAD1, 0xD8DFE6ED, 0xF4FB0209,
        0x10171E25, 0x2C333A41, 0x484F565D, 0x646B7279,
    ]
    SBOX = bytes([
        0xD6,0x90,0xE9,0xFE,0xCC,0xE1,0x3D,0xB7,0x16,0xB6,0x14,0xC2,0x28,0xFB,0x2C,0x05,
        0x2B,0x67,0x9A,0x76,0x2A,0xBE,0x04,0xC3,0xAA,0x44,0x13,0x26,0x49,0x86,0x06,0x99,
        0x9C,0x42,0x50,0xF4,0x91,0xEF,0x98,0x7A,0x33,0x54,0x0B,0x43,0xED,0xCF,0xAC,0x62,
        0xE4,0xB3,0x1C,0xA9,0xC9,0x08,0xE8,0x95,0x80,0xDF,0x94,0xFA,0x75,0x8F,0x3F,0xA6,
        0x47,0x07,0xA7,0xFC,0xF3,0x73,0x17,0xBA,0x83,0x59,0x3C,0x19,0xE6,0x85,0x4F,0xA8,
        0x68,0x6B,0x81,0xB2,0x71,0x64,0xDA,0x8B,0xF8,0xEB,0x0F,0x4B,0x70,0x56,0x9D,0x35,
        0x1E,0x24,0x0E,0x5E,0x63,0x58,0xD1,0xA2,0x25,0x22,0x7C,0x3B,0x01,0x21,0x78,0x87,
        0xD4,0x00,0x46,0x57,0x9F,0xD3,0x27,0x52,0x4C,0x36,0x02,0xE7,0xA0,0xC4,0xC8,0x9E,
        0xEA,0xBF,0x8A,0xD2,0x40,0xC7,0x38,0xB5,0xA3,0xF7,0xF2,0xCE,0xF9,0x61,0x15,0xA1,
        0xE0,0xAE,0x5D,0xA4,0x9B,0x34,0x1A,0x55,0xAD,0x93,0x32,0x30,0xF5,0x8C,0xB1,0xE3,
        0x1D,0xF6,0xE2,0x2E,0x82,0x66,0xCA,0x60,0xC0,0x29,0x23,0xAB,0x0D,0x53,0x4E,0x6F,
        0xD5,0xDB,0x37,0x45,0xDE,0xFD,0x8E,0x2F,0x03,0xFF,0x6A,0x72,0x6D,0x6C,0x5B,0x51,
        0x8D,0x1B,0xAF,0x92,0xBB,0xDD,0xBC,0x7F,0x11,0xD9,0x5C,0x41,0x1F,0x10,0x5A,0xD8,
        0x0A,0xC1,0x31,0x88,0xA5,0xCD,0x7B,0xBD,0x2D,0x74,0xD0,0x12,0xB8,0xE5,0xB4,0xB0,
        0x89,0x69,0x97,0x4A,0x0C,0x96,0x77,0x7E,0x65,0xB9,0xF1,0x09,0xC5,0x6E,0xC6,0x84,
        0x18,0xF0,0x7D,0xEC,0x3A,0xDC,0x4D,0x20,0x79,0xEE,0x5F,0x3E,0xD7,0xCB,0x39,0x48,
    ])

    @classmethod
    def _tau(cls, a: int) -> int:
        b = 0
        for i in range(4):
            b |= cls.SBOX[(a >> (8 * (3 - i))) & 0xFF] << (8 * (3 - i))
        return b

    @classmethod
    def _rotl(cls, x: int, n: int) -> int:
        return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF

    @classmethod
    def _l(cls, a: int) -> int:
        return a ^ cls._rotl(a, 2) ^ cls._rotl(a, 10) ^ cls._rotl(a, 18) ^ cls._rotl(a, 24)

    @classmethod
    def _lp(cls, a: int) -> int:
        return a ^ cls._rotl(a, 13) ^ cls._rotl(a, 23)

    @classmethod
    def _expand_key(cls, key: bytes) -> list:
        if len(key) != 16:
            raise ValueError("SM4 密钥长度必须为 16 字节")
        mk = [int.from_bytes(key[i:i + 4], "big") for i in range(0, 16, 4)]
        k = [mk[i] ^ cls.FK[i] for i in range(4)]
        rk = []
        for i in range(32):
            k[i % 4] ^= cls._lp(cls._tau(k[(i + 1) % 4] ^ k[(i + 2) % 4] ^ k[(i + 3) % 4] ^ cls.CK[i]))
            rk.append(k[i % 4])
        return rk

    @classmethod
    def _f(cls, x0: int, x1: int, x2: int, x3: int, rk: int) -> int:
        return x0 ^ cls._l(cls._tau(x1 ^ x2 ^ x3 ^ rk))

    @classmethod
    def _crypt_block(cls, block: bytes, rk: list) -> bytes:
        x = [int.from_bytes(block[i:i + 4], "big") for i in range(0, 16, 4)]
        for i in range(32):
            x.append(cls._f(x[i], x[i + 1], x[i + 2], x[i + 3], rk[i]))
        return x[35].to_bytes(4, "big") + x[34].to_bytes(4, "big") + \
               x[33].to_bytes(4, "big") + x[32].to_bytes(4, "big")

    @classmethod
    def _pad(cls, data: bytes) -> bytes:
        n = 16 - (len(data) % 16)
        return data + bytes([n] * n)

    @classmethod
    def _unpad(cls, data: bytes) -> bytes:
        return data[:-data[-1]]

    @classmethod
    def encrypt(cls, plaintext: bytes, key: bytes) -> bytes:
        """SM4 ECB 加密"""
        rk = cls._expand_key(key)
        padded = cls._pad(plaintext)
        result = b""
        for i in range(0, len(padded), 16):
            result += cls._crypt_block(padded[i:i + 16], rk)
        return result

    @classmethod
    def decrypt(cls, ciphertext: bytes, key: bytes) -> bytes:
        """SM4 ECB 解密"""
        rk = cls._expand_key(key)[::-1]
        result = b""
        for i in range(0, len(ciphertext), 16):
            result += cls._crypt_block(ciphertext[i:i + 16], rk)
        return cls._unpad(result)

    @classmethod
    def generate_key(cls) -> bytes:
        """生成 16 字节 SM4 密钥"""
        return os.urandom(16)


# ════════════════════════════════════════════════════
# SM3 哈希链 — 防篡改追溯
# ════════════════════════════════════════════════════

class SM3HashChain:
    """SM3 哈希链：每笔操作链接前一笔哈希，形成不可篡改的追溯链。

    用法:
        chain = SM3HashChain()
        chain.add("第一次操作的数据")
        chain.add("第二次操作的数据")
        chain.verify()  # → True
    """

    def __init__(self, genesis: Optional[bytes] = None):
        self.chain: List[dict] = []
        self._genesis = genesis or os.urandom(32)
        self._add_link("genesis", self._genesis, prev_hash=bytes(32))

    def _add_link(self, label: str, data: bytes, prev_hash: bytes):
        ts = time.time()
        link_hash = SM3.hash(prev_hash + data + struct.pack(">d", ts))
        self.chain.append({
            "index": len(self.chain),
            "label": label,
            "timestamp": ts,
            "data_hash": SM3.hex(data),
            "prev_hash": prev_hash.hex() if prev_hash != bytes(32) else "0" * 64,
            "link_hash": link_hash.hex(),
        })
        return link_hash

    def add(self, label: str, data: Union[str, bytes]) -> str:
        """添加一个节点，返回 link_hash"""
        if isinstance(data, str):
            data = data.encode("utf-8")
        prev = bytes.fromhex(self.chain[-1]["link_hash"]) if self.chain else bytes(32)
        return self._add_link(label, data, prev).hex()

    def verify(self) -> tuple[bool, List[int]]:
        """验证全链完整性 → (是否通过, 失败节点索引列表)"""
        failures = []
        for i in range(1, len(self.chain)):
            link = self.chain[i]
            prev = self.chain[i - 1]
            expected_prev = prev["link_hash"]
            if link["prev_hash"] != expected_prev:
                failures.append(i)
        return (len(failures) == 0, failures)

    def last_hash(self) -> str:
        """当前链尾哈希"""
        return self.chain[-1]["link_hash"]

    def to_list(self) -> List[dict]:
        """导出为 JSON 可序列化列表"""
        return self.chain


# ════════════════════════════════════════════════════
# 自检
# ════════════════════════════════════════════════════

if __name__ == "__main__":
    # SM3 自检
    expected = "66c7f0f462eeedd9d1f2d46bdc10e4e24167c4875cf2f7a2297da02b8f4ba8e0"
    assert SM3.hex("abc") == expected, "SM3 自检失败"
    print("🟢 SM3 自检通过")

    # SM4 自检
    key = b"0123456789abcdef"
    pt = b"hello CNSH guomi!"
    ct = SM4.encrypt(pt, key)
    dt = SM4.decrypt(ct, key)
    assert dt == pt, "SM4 自检失败"
    print("🟢 SM4 加解密自检通过")

    # 哈希链自检
    chain = SM3HashChain()
    chain.add("op1", "data one")
    chain.add("op2", "data two")
    ok, fails = chain.verify()
    assert ok, "哈希链自检失败"
    print("🟢 SM3 哈希链自检通过")
    print("🟢 全部国密模块自检通过")

File: web/longhun-guardian-dashboard/api/lib/sm-crypto.ts
# 归属名: 诸葛鑫 | UID9622 · 龍芯北辰
/**
 * 龍魂国密算法库 v1.0
 * DNA: #龍芯⚡️丙午·乙未·丙戌·甲午·䷕贲-LONGHUN-SM-CRYPTO-v1.0
 * 实现: SM3哈希 / SM4对称加密 / DNA签名
 * 无外部依赖，纯TypeScript实现
 */

// ========== SM3 国密哈希 ==========

const SM3_IV = new Uint32Array([
  0x7380166f, 0x4914b2b9, 0x172442d7, 0xda8a0600,
  0xa96f30bc, 0x163138aa, 0xe38dee4d, 0xb0fb0e4e,
]);

function rotateLeft32(x: number, n: number): number {
  return (x << n) | (x >>> (32 - n));
}

function ff(x: number, y: number, z: number, j: number): number {
  return j < 16 ? x ^ y ^ z : (x & y) | (x & z) | (y & z);
}

function gg(x: number, y: number, z: number, j: number): number {
  return j < 16 ? x ^ y ^ z : (x & y) | (~x & z);
}

function p0(x: number): number {
  return x ^ rotateLeft32(x, 9) ^ rotateLeft32(x, 17);
}

function p1(x: number): number {
  return x ^ rotateLeft32(x, 15) ^ rotateLeft32(x, 23);
}

function sm3Compress(h: Uint32Array, block: Uint8Array): void {
  const w = new Uint32Array(68);
  const w1 = new Uint32Array(64);

  for (let i = 0; i < 16; i++) {
    w[i] =
      (block[i * 4] << 24) |
      (block[i * 4 + 1] << 16) |
      (block[i * 4 + 2] << 8) |
      block[i * 4 + 3];
  }

  for (let i = 16; i < 68; i++) {
    w[i] = p1(w[i - 16] ^ w[i - 9] ^ rotateLeft32(w[i - 3], 15)) ^
      rotateLeft32(w[i - 13], 7) ^
      w[i - 6];
  }

  for (let i = 0; i < 64; i++) {
    w1[i] = w[i] ^ w[i + 4];
  }

  const t = [0x79cc4519, 0x7a879d8a];
  let [a, b, c, d, e, f, g, h0] = [h[0], h[1], h[2], h[3], h[4], h[5], h[6], h[7]];

  for (let j = 0; j < 64; j++) {
    const ss1 = rotateLeft32(
      rotateLeft32(a, 12) + e + rotateLeft32(t[j < 16 ? 0 : 1], j % 32),
      7,
    );
    const ss2 = ss1 ^ rotateLeft32(a, 12);
    const tt1 = ff(a, b, c, j) + d + ss2 + w1[j];
    const tt2 = gg(e, f, g, j) + h0 + ss1 + w[j];
    d = c;
    c = rotateLeft32(b, 9);
    b = a;
    a = tt1;
    h0 = g;
    g = rotateLeft32(f, 19);
    f = e;
    e = p0(tt2);
  }

  h[0] ^= a; h[1] ^= b; h[2] ^= c; h[3] ^= d;
  h[4] ^= e; h[5] ^= f; h[6] ^= g; h[7] ^= h0;
}

export function sm3(message: string | Uint8Array): string {
  const data = typeof message === "string"
    ? new TextEncoder().encode(message)
    : message;

  const bitLen = BigInt(data.length) * 8n;
  const padLen = (64 - ((data.length + 9) % 64)) % 64;
  const totalLen = data.length + 1 + padLen + 8;
  const buf = new Uint8Array(totalLen);

  buf.set(data);
  buf[data.length] = 0x80;

  const dv = new DataView(buf.buffer, totalLen - 8);
  dv.setUint32(0, Number(bitLen >> 32n), false);
  dv.setUint32(4, Number(bitLen & 0xffffffffn), false);

  const h = new Uint32Array(SM3_IV);

  for (let i = 0; i < totalLen; i += 64) {
    sm3Compress(h, buf.subarray(i, i + 64));
  }

  return Array.from(h).map((x) => x.toString(16).padStart(8, "0")).join("");
}

// ========== SM4 国密对称加密 ==========

const SM4_SBOX = new Uint8Array([
  0xd6, 0x90, 0xe9, 0xfe, 0xcc, 0xe1, 0x3d, 0xb7, 0x16, 0xb6, 0x14, 0xc2, 0x28,
  0xfb, 0x2c, 0x05, 0x2b, 0x67, 0x9a, 0x76, 0x2a, 0xbe, 0x04, 0xc3, 0xaa, 0x44,
  0x13, 0x26, 0x49, 0x86, 0x06, 0x99, 0x9c, 0x42, 0x50, 0xf4, 0x91, 0xef, 0x98,
  0x7a, 0x33, 0x54, 0x0b, 0x43, 0xed, 0xcf, 0xac, 0x62, 0xe4, 0xb3, 0x1c, 0xa9,
  0xc9, 0x08, 0xe8, 0x95, 0x80, 0xdf, 0x94, 0xfa, 0x75, 0x8f, 0x3f, 0xa6, 0x47,
  0x07, 0xa7, 0xfc, 0xf3, 0x73, 0x17, 0xba, 0x83, 0x59, 0x3c, 0x19, 0xe6, 0x85,
  0x4f, 0xa8, 0x68, 0x6b, 0x81, 0xb2, 0x71, 0x64, 0xda, 0x8b, 0xf8, 0xeb, 0x0f,
  0x4b, 0x70, 0x56, 0x9d, 0x35, 0x1e, 0x24, 0x0e, 0x5e, 0x63, 0x58, 0xd1, 0xa2,
  0x25, 0x22, 0x7c, 0x3b, 0x01, 0x21, 0x78, 0x87, 0xd4, 0x00, 0x46, 0x57, 0x9f,
  0xd3, 0x27, 0x52, 0x4c, 0x36, 0x02, 0xe7, 0xa0, 0xc4, 0xc8, 0x9e, 0xea, 0xbf,
  0x8a, 0xd2, 0x40, 0xc7, 0x38, 0xb5, 0xa3, 0xf7, 0xf2, 0xce, 0xf9, 0x61, 0x15,
  0xa1, 0xe0, 0xae, 0x5d, 0xa4, 0x9b, 0x34, 0x1a, 0x55, 0xad, 0x93, 0x32, 0x30,
  0xf5, 0x8c, 0xb1, 0xe3, 0x1d, 0xf6, 0xe2, 0x2e, 0x82, 0x66, 0xca, 0x60, 0xc0,
  0x29, 0x23, 0xab, 0x0d, 0x53, 0x4e, 0x6f, 0xd5, 0xdb, 0x37, 0x45, 0xde, 0xfd,
  0x8e, 0x2f, 0x03, 0xff, 0x6a, 0x72, 0x6d, 0x6c, 0x5b, 0x51, 0x8d, 0x1b, 0xaf,
  0x92, 0xbb, 0xdd, 0xbc, 0x7f, 0x11, 0xd9, 0x5c, 0x41, 0x1f, 0x10, 0x5a, 0xd8,
  0x0a, 0xc1, 0x31, 0x88, 0xa5, 0xcd, 0x7b, 0xbd, 0x2d, 0x74, 0xd0, 0x12, 0xb8,
  0xe5, 0xb4, 0xb0, 0x89, 0x69, 0x97, 0x4a, 0x0c, 0x96, 0x77, 0x7e, 0x65, 0xb9,
  0xf1, 0x09, 0xc5, 0x6e, 0xc6, 0x84, 0x18, 0xf0, 0x7d, 0xec, 0x3a, 0xdc, 0x4d,
  0x20, 0x79, 0xee, 0x5f, 0x3e, 0xd7, 0xcb, 0x39, 0x48,
]);

const SM4_FK = new Uint32Array([0xa3b1bac6, 0x56aa3350, 0x677d9197, 0xb27022dc]);
const SM4_CK = new Uint32Array([
  0x00070e15, 0x1c232a31, 0x383f464d, 0x545b6269, 0x70777e85, 0x8c939aa1,
  0xa8afb6bd, 0xc4cbd2d9, 0xe0e7eef5, 0xfc030a11, 0x181f262d, 0x343b4249,
  0x50575e65, 0x6c737a81, 0x888f969d, 0xa4abb2b9, 0xc0c7ced5, 0xdce3eaf1,
  0xf8ff060d, 0x141b2229, 0x30373e45, 0x4c535a61, 0x686f767d, 0x848b9299,
  0xa0a7aeb5, 0xbcc3cad1, 0xd8dfe6ed, 0xf4fb0209, 0x10171e25, 0x2c333a41,
  0x484f565d, 0x646b7279,
]);

function sm4_tau(a: number): number {
  return (
    (SM4_SBOX[a >>> 24] << 24) |
    (SM4_SBOX[(a >>> 16) & 0xff] << 16) |
    (SM4_SBOX[(a >>> 8) & 0xff] << 8) |
    SM4_SBOX[a & 0xff]
  );
}

function sm4_l(b: number): number {
  return (
    b ^
    ((b << 2) | (b >>> 30)) ^
    ((b << 10) | (b >>> 22)) ^
    ((b << 18) | (b >>> 14)) ^
    ((b << 24) | (b >>> 8))
  );
}

function sm4_f(x0: number, x1: number, x2: number, x3: number, rk: number): number {
  return x0 ^ sm4_l(sm4_tau(x1 ^ x2 ^ x3 ^ rk));
}

function sm4_key_expansion(key: Uint8Array): Uint32Array {
  const rk = new Uint32Array(32);
  const k = new Uint32Array(4);

  for (let i = 0; i < 4; i++) {
    k[i] =
      (key[i * 4] << 24) |
      (key[i * 4 + 1] << 16) |
      (key[i * 4 + 2] << 8) |
      key[i * 4 + 3];
    k[i] ^= SM4_FK[i];
  }

  for (let i = 0; i < 32; i++) {
    k[i % 4] ^= sm4_f(k[0], k[1], k[2], k[3], SM4_CK[i]);
    rk[i] = k[i % 4];
  }

  return rk;
}

function sm4_crypt_block(input: Uint8Array, output: Uint8Array, rk: Uint32Array): void {
  const x = new Uint32Array(4);
  for (let i = 0; i < 4; i++) {
    x[i] =
      (input[i * 4] << 24) |
      (input[i * 4 + 1] << 16) |
      (input[i * 4 + 2] << 8) |
      input[i * 4 + 3];
  }

  const tmp = new Uint32Array(36);
  tmp.set(x);

  for (let i = 0; i < 32; i++) {
    tmp[i + 4] = sm4_f(tmp[i], tmp[i + 1], tmp[i + 2], tmp[i + 3], rk[i]);
  }

  const out = new Uint32Array(4);
  out[0] = tmp[35]; out[1] = tmp[34]; out[2] = tmp[33]; out[3] = tmp[32];

  for (let i = 0; i < 4; i++) {
    output[i * 4] = (out[i] >>> 24) & 0xff;
    output[i * 4 + 1] = (out[i] >>> 16) & 0xff;
    output[i * 4 + 2] = (out[i] >>> 8) & 0xff;
    output[i * 4 + 3] = out[i] & 0xff;
  }
}

function pkcs7_pad(data: Uint8Array, blockSize: number = 16): Uint8Array {
  const padLen = blockSize - (data.length % blockSize);
  const result = new Uint8Array(data.length + padLen);
  result.set(data);
  result.fill(padLen, data.length);
  return result;
}

function pkcs7_unpad(data: Uint8Array): Uint8Array {
  const padLen = data[data.length - 1];
  return data.subarray(0, data.length - padLen);
}

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}

export function sm4_encrypt(plaintext: string, keyHex: string): string {
  const key = hexToBytes(keyHex.length === 32 ? keyHex : sm3(keyHex).substring(0, 32));
  const data = new TextEncoder().encode(plaintext);
  const padded = pkcs7_pad(data);
  const rk = sm4_key_expansion(key);
  const output = new Uint8Array(padded.length);

  for (let i = 0; i < padded.length; i += 16) {
    sm4_crypt_block(padded.subarray(i, i + 16), output.subarray(i, i + 16), rk);
  }

  return bytesToHex(output);
}

export function sm4_decrypt(cipherHex: string, keyHex: string): string {
  const key = hexToBytes(keyHex.length === 32 ? keyHex : sm3(keyHex).substring(0, 32));
  const data = hexToBytes(cipherHex);
  const rk = sm4_key_expansion(key);
  const output = new Uint8Array(data.length);

  for (let i = 0; i < data.length; i += 16) {
    sm4_crypt_block(data.subarray(i, i + 16), output.subarray(i, i + 16), rk);
  }

  const unpadded = pkcs7_unpad(output);
  return new TextDecoder().decode(unpadded);
}

// ========== DNA 签名 ==========

export function dnaSign(data: string, secret: string): string {
  const hash1 = sm3(secret + data);
  const hash2 = sm3(hash1 + "龍魂");
  const hash3 = sm3(hash2 + secret.substring(0, Math.min(secret.length, 16)));
  const timestamp = new Date().toISOString().split("T")[0];
  const hash4 = sm3(hash3 + timestamp);
  return `#龍芯⚡️${timestamp}-LONGHUN-SIGN-${hash4.substring(0, 16).toUpperCase()}`;
}

export function dnaVerify(data: string, secret: string, signature: string): boolean {
  return signature === dnaSign(data, secret);
}

export function generateSM4Key(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return bytesToHex(bytes);
}

export function generateSecretKey(): string {
  const bytes = new Uint8Array(32);
  crypto.getRandomValues(bytes);
  return bytesToHex(bytes);
}
